python/observation-publisher/publisher/prediction_log_consumer.py

```diff
@@ -0,0 +1,91 @@
+from typing import List, Optional, Union
+
+import numpy as np
+import pandas as pd
+from caraml.upi.v1.prediction_log_pb2 import PredictionLog
+from confluent_kafka import Consumer, KafkaException
+from google.protobuf.internal.well_known_types import ListValue
+
+from publisher.config import (KafkaConsumerConfig, ModelSchema, ModelSpec,
+                              ValueType)
+from publisher.observability_backend import ObservationSink
+
+
+def parse_message_to_prediction_log(msg: str) -> PredictionLog:
+    log = PredictionLog()
+    log.ParseFromString(msg)
+    return log
+
+
+def poll_prediction_logs(
+    kafka_consumer: Consumer,
+    consumer_cfg: KafkaConsumerConfig,
+    observation_sink: ObservationSink,
+    model_cfg: ModelSpec,
+):
+    kafka_consumer.subscribe([consumer_cfg.topic])
+    while True:
+        messages = kafka_consumer.consume(
+            consumer_cfg.batch_size, timeout=consumer_cfg.poll_timeout_seconds
+        )
+        errors = [msg.error() for msg in messages if msg.error() is not None]
+
+        if len(errors) > 0:
+            print(f"Last encountered error: {errors[-1]}")
+            raise KafkaException(errors[-1])
+        prediction_logs = [
+            parse_message_to_prediction_log(msg.value().decode("utf-8"))
+            for msg in messages
+            if (msg is not None and msg.error() is None)
+        ]
+        if len(prediction_logs) > 0:
+            df = log_to_dataframe(prediction_logs, model_cfg.schema)
+            observation_sink.write(dataframe=df)
```

Comment:
this can potentially cause lagging, can we publish it async?

Reply:
We could publish it async, but that comes at the risk of data loss. For example, if a commit has been performed before the write is completed, all the messages will be lost forever should the write ultimately failed. Unlike producer, there are a lot more messages in a single polling request (especially if we want to fulfil the min batch size of 50Mb per poll).

Reply:
Yes, it will be the same set of messages.